jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,420 @@
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = (HashSet<Integer>) this.labels.clone();
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    /**
+     * Implementation of a key for the path maps.
+     * It is used in AhujaOrlinSharmaCyclicExchangeLocalAugmentation to efficiently maintain the path sets in the calculation.
+     *
+     * @param <V> the vertex type
+     *
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class PathSetKey<V> {
+        /**
+         * the head of the paths indexed by this key
+         */
+        private V head;
+        /**
+         * the tail of the paths indexed by this key
+         */
+        private V tail;
+        /**
+         * the label set of the paths indexed by this key
+         */
+        private Set<Integer> labels;
+
+        /**
+         * Constructs a new PathSetKey object
+         *
+         * @param head the head of the paths indexed by this key
+         * @param tail the tail of the paths indexed by this key
+         * @param labels the label set of the paths indexed by this key
+         */
+        private PathSetKey(V head, V tail, Set<Integer> labels) {
+            this.head = head;
+            this.tail = tail;
+            this.labels = labels;
+        }
+
+        /**
+         * Constructs a new PathSetKey object
+         *
+         * @param head the head of the path indexed by this key
+         * @param tail the tail of the path indexed by this key
+         * @param labels the label set of the path indexed by this key
+         * @param <V> the vertex type
+         * @return a new PathSetKey object
+         */
+        protected static <V> PathSetKey<V> of(V head, V tail, Set<Integer> labels) {
+            return new PathSetKey<>(head, tail, labels);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(this.head, this.tail, this.labels);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            else if (!(o instanceof PathSetKey))
+                return false;
+
+            @SuppressWarnings("unchecked") PathSetKey<V> other = (PathSetKey<V>) o;
+            return Objects.equals(head, other.head) && Objects.equals(tail, other.tail) && Objects.equals(labels, other.labels);
+        }
+    }
+
+    /**
+     * the input graph
+     */
+    private Graph<V, E> graph;
```

Comment:
These fields should be at the beginning of the class, follows by the public constructors.